include/xpedite/common/LeasedPtr.H

```diff
@@ -0,0 +1,99 @@
+//////////////////////////////////////////////////////////////////////////////////////////////////////
+//
+// LeasedPtr - a SPSC wait free data structure for sharing a resource between two threads.
+//
+// It has a clear ownership semantics, allowing the consumer to lease the pointer from the producer.
+//
+// Once the consumer has successfully borrowed the resource, the producer will know not to deallocate
+// the resource until the consumer has returned the lease.
+//
+// If the consumer takes too long to return the resource, the producer can release the ownership.
+// The LeasedPtr will ensure the consumer knows it now owns the resource and is responsible for deallocation.
+//
+// Author: Marcin Dlugajczyk, Morgan Stanley
+//
+//////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+#include <atomic>
+
+namespace xpedite { namespace common {
+
+  namespace {
+    static constexpr char leased{};
+  }
+
+  template<typename T>
+  class LeasedPtr
+  {
+    std::atomic<T *> _ptr{nullptr};
+
+    // Returns a special pointer value used to indicate the pointer is currently leased.
+    constexpr T* getLeasedMarker() {
+      return const_cast<T*>(reinterpret_cast<const T*>(&leased));
+    }
+
+  public:
+
+    LeasedPtr() noexcept = default;
+    explicit LeasedPtr(T *ptr_) noexcept : _ptr{ptr_} {}
+    ~LeasedPtr() = default;
+    LeasedPtr<T>(LeasedPtr<T> &other) = delete;
+    LeasedPtr<T>(LeasedPtr<T> &&other) = delete;
+    LeasedPtr<T>& operator=(LeasedPtr<T> &other) = delete;
+    LeasedPtr<T>& operator=(LeasedPtr<T> &&other) = delete;
+
+    // Tries to deposit the `ptr_`, and make it available to the consumer.
+    // Returns true if the pointer was susscessfully deposited.
+    // Returns false if the LeasedPtr already contains a different pointer.
+    bool tryDeposit(T *ptr_) noexcept {
+      T *empty{nullptr};
+      return _ptr.compare_exchange_strong(empty, ptr_, std::memory_order_release, std::memory_order_relaxed);
+    }
+
+    // Indicates if has space for new deposit.
+    bool empty() noexcept {
+      return _ptr.load(std::memory_order_relaxed) == nullptr;
+    }
+
+    // Producer breaks the lease, and releases the ownership of the pointer and passes it to the consumer.
+    // Returns true on success. Returns false if the consumer hasn't leased the pointer yet - the ownership stays with the producer.
+    bool breakLease() noexcept {
```

Comment:
`breakLease()` can be renamed to `revoke()` to match `provision()`
This class is great, since it allows the ownership of the underlying object to be flexibly moved between leaser and leasee. However it currently put the onus of deletion of the underlying object on the callers.
However it would be better, if it also handled life time management.

For a leaseable object, there are potentially three possible unique owners at any point in time.
Consider the following sequence of events.
**time t1)** thread#1 creates a request object -> auto req = std::make_unique<Request>(...)
    Thread#1 is the unique owner of the request
**time t2)** thread#1 decides to lease the object out and calls 
    leaseablePtr.provision(std::move(req));
   For the time being the ownership is moved to LeasablePtr
**time t3. scenario - a)** thread#1 decides to break the lease
  leasablePtr.revoke() -> should return a unique_ptr {}.
  If the revoke() was called before the object is leased by any other thread, the object will be moved and ownership will be transferred back to thread#1 by the unique ptr.
  On the other hand, If the object is already leased, the ownership is with the leased thread. This should just return an empty unique_ptr.

**time t3.scenario - b)** thread#2 attempts to lease the object
  leasablePtr.lease() should return a unique_ptr, 
  If the lease succeeds, the ownership is transferred to the leased thread by the unique_ptr.
  If there is nothing to lease, the caller gets an empty ptr.

**time t4)** thread#2 attempts to return the leased object
  bool returnLease(std::unique_ptr<T> ..)
  If the thead#1 is still waiting, the lease can be returned. The parameter get's moved in and LeasablePtr gets the ownership.
  if the thead#1 broke the lease early, the unique_ptr can be reset() there by deleting the object being returned.

something like below
```
class LeasablePtr {
  enum class State {
    DORMANT, PROVISIONED, LEASED, REVOKED, RETURNED
  }; 
  State _state;
  std::unique_ptr<T> _ptr;

  // The object start at DORMANT state
  // call to provision(), moves object to _ptr and _state to PROVISIONED 
  // while PROVISIONED, a call to revoke() will move it to DORMANT and move _ptr to caller 
 // while PROVISIONED, a call to lease() will move it to LEASED and move _ptr to caller
 // while LEASED, a call to revoke() will move it to REVOKED and return empty ptr
 // while LEASED, a call to returnLease() will move parameter to _ptr and _state to RETURNED 
 // while in REVOKED state, a call to returnLease() will move it to DORMANT and delete the object
// while in RETURNED state, a call to revoke() will move it to DORMANT and move _ptr to caller.
The atomics will be done only on the _state, while _ptr can do object ownership, 
};
```